cmd/balena-engine/sched_linux.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"io/ioutil"
+	"math"
+	"strconv"
+	"time"
+	"unsafe"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+// Scheduling policies.
+const (
+	SCHED_NORMAL        = 0
+	SCHED_FIFO          = 1
+	SCHED_RR            = 2
+	SCHED_BATCH         = 3
+	SCHED_RESET_ON_FORK = 0x40000000 // Meant to be ORed with the others
+)
+
+// setScheduler sets the scheduling policy and priority of all threads in the
+// current process.
+func setScheduler(policy, prio int) {
+	// The Go runtime can (and does) create new threads after the program
+	// startup. We therefore keep looping for some minutes in a goroutine,
+	// setting the priorities of any new threads. From experimentation, we see
+	// no new threads created after a handful of minutes, so we don't need an
+	// infinite loop.
+	go func() {
```

Comment:
@lmbarros I was more thinking of a lower level implementation in balenad and containerd that would set scheduling policies at the point where threads are created. The implementation here is fine for an initial approach, but it does not guarantee setting the scheduling policy on all threads.
What's the reasoning for deciding on this type of implementation?

Reply:
@alexgg That's the best solution I could find so far. These threads are created by the Go runtime and it hides them completely from the Go programmer. The only way I can think of is changing the Go runtime itself, but I see this as a very last resort. (But your point is valid for sure, I share the same concern. That's probably worth more investigation.)